# Movies Tonight Part 3: SQL DML

In [1]:
%load_ext sql
%sql sqlite:///Movies.db

'Connected: @Movies.db'

1. **Load the `DATASET.csv` file into the `Movies.db` database.**  
    - Import the CSV file to create a new table called `DATASET`. Follow the example [here](http://www.sqlitetutorial.net/sqlite-import-csv/).
    - Use a [`PRAGMA table-info` pseudo-query](https://www.sqlite.org/pragma.html#pragma_table_info) to inspect the design of the new table. [How are we going to handle dates](https://www.sqlite.org/datatype3.html)?

In [6]:
%%sql

#Special select query for metadata
PRAGMA table_info(DATASET);

#Update ShowTime to date
UPDATE DATASET

strftime('%H:%M')

 * sqlite:///Movies.db
Done.
Done.


[(0, 'TName', 'TEXT', 0, None, 0),
 (1, 'Location', 'TEXT', 0, None, 0),
 (2, 'Phone', 'TEXT', 0, None, 0),
 (3, 'MTitle', 'TEXT', 0, None, 0),
 (4, 'ShowTime', 'TEXT', 0, None, 0),
 (5, 'Rating', 'TEXT', 0, None, 0),
 (6, 'CCode', 'TEXT', 0, None, 0),
 (7, 'CName', 'TEXT', 0, None, 0)]

2. **Populate the tables with data.**  

    - Use `INSERT ... SELECT ...` queries to extract data from `DATASET` for each table of your other tables. 
    - Start with the strong entities before moving on to the weak ones. Be sure to always select distinct     values. 
    - Foreign Keys get special treatment. You will have to use JOINs to pick up the PKs from the other side of the relationships. (As in Part 1, use the fact that movie names, theater names, and people names are always unique in our source data. That way you can use then for the JOIN columns. 
    - If you have to start over, then rerun your DDL queries from Part 2. 

In [38]:
%%sql
DELETE FROM MOVIE;
INSERT INTO MOVIE (Title,Rating)
    SELECT DISTINCT MTitle, Rating 
        FROM DATASET;
    SELECT * FROM MOVIE;

DELETE FROM THEATER;
INSERT INTO THEATER (TheaterName,Location,Phone)
    SELECT DISTINCT TName, Location, Phone
        FROM DATASET;
    SELECT * FROM THEATER;

DELETE FROM ARTIST;
INSERT INTO ARTIST (ArtistName)
    SELECT DISTINCT CName
        FROM DATASET;
        SELECT * FROM ARTIST;
        
DELETE FROM SHOWTIME;
INSERT INTO SHOWTIME (ShowTime,MovieID,TheaterID)
SELECT DISTINCT ShowTime, MOVIE.MovieID, THEATER.TheaterID
FROM DATASET
    JOIN MOVIE ON (MOVIE.Title = DATASET.MTitle)
    JOIN THEATER ON (THEATER.TheaterName = DATASET.TName);

DELETE FROM CREDIT;
INSERT INTO CREDIT (CCode,ArtistID,ArtistName,MovieID,Title)
SELECT DISTINCT CCode, ARTIST.ArtistID, ARTIST.ArtistName, MOVIE.MovieID, MOVIE.Title
FROM DATASET
    JOIN ARTIST ON (ARTIST.ArtistName = DATASET.CName)
    JOIN MOVIE ON (MOVIE.Title = DATASET.MTitle);
SELECT * FROM CREDIT;






 * sqlite:///Movies.db
23 rows affected.
23 rows affected.
Done.
9 rows affected.
9 rows affected.
Done.
153 rows affected.
153 rows affected.
Done.
131 rows affected.
131 rows affected.
162 rows affected.
162 rows affected.
Done.


[(1, 'A', 1, 'Austin Pendleton', 1, 'Associate, The'),
 (2, 'A', 2, 'Bebe Neuwirth', 1, 'Associate, The'),
 (3, 'A', 3, 'Dianne Wiest', 1, 'Associate, The'),
 (4, 'A', 4, 'Eli Wallach', 1, 'Associate, The'),
 (5, 'A', 5, 'Kenny Kerr', 1, 'Associate, The'),
 (6, 'A', 6, 'Lainie Kazan', 1, 'Associate, The'),
 (7, 'A', 7, 'Tim Daly', 1, 'Associate, The'),
 (8, 'A', 8, 'Whoopi Goldberg', 1, 'Associate, The'),
 (9, 'D', 9, 'Donald Petrie', 1, 'Associate, The'),
 (10, 'A', 10, 'Bernard Hill', 2, 'Ghost & The Darkness, The'),
 (11, 'A', 11, 'Brian McCardie', 2, 'Ghost & The Darkness, The'),
 (12, 'A', 12, 'Emily Mortimer', 2, 'Ghost & The Darkness, The'),
 (13, 'A', 13, 'Henry Cele', 2, 'Ghost & The Darkness, The'),
 (14, 'A', 14, 'John Kani', 2, 'Ghost & The Darkness, The'),
 (15, 'A', 15, 'Michael Douglas', 2, 'Ghost & The Darkness, The'),
 (16, 'A', 16, 'Tom Wilkinson', 2, 'Ghost & The Darkness, The'),
 (17, 'A', 17, 'Val Kilmer', 2, 'Ghost & The Darkness, The'),
 (18, 'D', 18, 'Stephen Hopkins', 2, 'Ghost & The Darkness, The'),
 (19, 'A', 19, 'Bill Pullman', 3, 'Independence Day'),
 (20, 'A', 20, 'Bill Smitrovich', 3, 'Independence Day'),
 (21, 'A', 21, 'Harvey Fierstein', 3, 'Independence Day'),
 (22, 'A', 22, 'Jeff Goldblum', 3, 'Independence Day'),
 (23, 'A', 23, 'Judd Hirsch', 3, 'Independence Day'),
 (24, 'A', 24, 'Margaret Colin', 3, 'Independence Day'),
 (25, 'A', 25, 'Mary McDonnell', 3, 'Independence Day'),
 (26, 'A', 26, 'Randy Quaid', 3, 'Independence Day'),
 (27, 'A', 27, 'Robert Loggia', 3, 'Independence Day'),
 (28, 'A', 28, 'Vivica Fox', 3, 'Independence Day'),
 (29, 'A', 29, 'Will Smith', 3, 'Independence Day'),
 (30, 'D', 30, 'Roland Emmerich', 3, 'Independence Day'),
 (31, 'A', 31, 'Emilio Estevez', 4, 'D3: The Mighty Ducks'),
 (32, 'A', 32, 'Heidi Kling', 4, 'D3: The Mighty Ducks'),
 (33, 'A', 33, 'Jeffrey Nordling', 4, 'D3: The Mighty Ducks'),
 (34, 'A', 34, 'Joss Ackland', 4, 'D3: The Mighty Ducks'),
 (35, 'A', 35, 'Rob Lieberman', 4, 'D3: The Mighty Ducks'),
 (36, 'D', 35, 'Rob Lieberman', 4, 'D3: The Mighty Ducks'),
 (37, 'A', 36, 'Greg Kinnear', 5, 'Dear God'),
 (38, 'A', 37, 'Hector Elizondo', 5, 'Dear God'),
 (39, 'A', 38, 'Jon Seda', 5, 'Dear God'),
 (40, 'A', 39, 'Laurie Metcalf', 5, 'Dear God'),
 (41, 'A', 40, 'Maria Pitillo', 5, 'Dear God'),
 (42, 'A', 41, 'Roscoe Lee Browne', 5, 'Dear God'),
 (43, 'A', 42, 'Tim Conway', 5, 'Dear God'),
 (44, 'D', 43, 'Garry Marshall', 5, 'Dear God'),
 (45, 'A', 44, 'Bette Midler', 6, 'First Wives Club, The'),
 (46, 'A', 45, 'Diane Keaton', 6, 'First Wives Club, The'),
 (47, 'A', 46, 'Goldie Hawn', 6, 'First Wives Club, The'),
 (48, 'D', 47, 'Hugh Wilson', 6, 'First Wives Club, The'),
 (49, 'A', 48, 'and Louise Fletcher', 7, 'High School High'),
 (50, 'A', 49, 'Jon Lovitz', 7, 'High School High'),
 (51, 'A', 50, 'Mekhi Phifer', 7, 'High School High'),
 (52, 'A', 51, 'Tia Carrera', 7, 'High School High'),
 (53, 'D', 52, 'Hart Bochner', 7, 'High School High'),
 (54, 'A', 53, 'Bill Murray', 8, 'Larger Than Life'),
 (55, 'A', 54, 'Janeane Garofalo', 8, 'Larger Than Life'),
 (56, 'A', 55, 'Jeremy Piven', 8, 'Larger Than Life'),
 (57, 'A', 56, 'Keith David', 8, 'Larger Than Life'),
 (58, 'A', 57, 'Matthew McConaughey', 8, 'Larger Than Life'),
 (59, 'A', 58, 'Pat Hingle', 8, 'Larger Than Life'),
 (60, 'D', 59, 'Howard Franklin', 8, 'Larger Than Life'),
 (61, 'A', 60, 'Barbra Streisand', 9, 'Mirror Has Two Faces, The'),
 (62, 'A', 61, 'Jeff Bridges', 9, 'Mirror Has Two Faces, The'),
 (63, 'A', 62, 'Lauren Bacall', 9, 'Mirror Has Two Faces, The'),
 (64, 'A', 63, 'Pierce Brosnan', 9, 'Mirror Has Two Faces, The'),
 (65, 'D', 60, 'Barbra Streisand', 9, 'Mirror Has Two Faces, The'),
 (66, 'A', 64, 'Delroy Lindo', 10, 'Ransom'),
 (67, 'A', 65, 'Gary Sinese', 10, 'Ransom'),
 (68, 'A', 66, 'Lili Taylor', 10, 'Ransom'),
 (69, 'A', 67, 'Mel Gibson', 10, 'Ransom'),
 (70, 'A', 68, 'Rene Russo', 10, 'Ransom'),
 (71, 'D', 69, 'Ron Howard', 10, 'Ransom'),
 (72, 'A', 70, 'Billy Crudup', 11,

3. **Check your work.**  
There are 9 theaters, 23 movies, 152 artists, 131 shows, and 161 credits.

In [ ]:
YOUR CODE HERE

4. **Now check the source data.**  
There are actually 13 duplicate rows in the original source data! Can you spot them? Write a SQL query that counts how many times each (MTitle, TName, Showtime, CCode, and CName) appears. Which ones appear more than once?

In [41]:
%%sql
SELECT MTitle, TName, ShowTime, CCode, CName, COUNT(*)
FROM DATASET
GROUP BY MTitle, TName, ShowTime, CCode, CName
HAVING COUNT(*) > 1;

 * sqlite:///Movies.db
Done.


[('Dear God', 'SOCAL Canyon Springs Cinema', '9:00 PM', 'A', 'Greg Kinnear', 2),
 ('Dear God', 'SOCAL Canyon Springs Cinema', '9:00 PM', 'A', 'Hector Elizondo', 2),
 ('Dear God', 'SOCAL Canyon Springs Cinema', '9:00 PM', 'A', 'Jon Seda', 2),
 ('Dear God', 'SOCAL Canyon Springs Cinema', '9:00 PM', 'A', 'Laurie Metcalf', 2),
 ('Dear God', 'SOCAL Canyon Springs Cinema', '9:00 PM', 'A', 'Maria Pitillo', 2),
 ('Dear God', 'SOCAL Canyon Springs Cinema', '9:00 PM', 'A', 'Roscoe Lee Browne', 2),
 ('Dear God', 'SOCAL Canyon Springs Cinema', '9:00 PM', 'A', 'Tim Conway', 2),
 ('Dear God', 'SOCAL Canyon Springs Cinema', '9:00 PM', 'D', 'Garry Marshall', 2),
 ('High School High', 'United Artists Riverside (Galleria) Tyler Mall', '9:00 PM', 'A', 'Jon Lovitz', 2),
 ('High School High', 'United Artists Riverside (Galleria) Tyler Mall', '9:00 PM', 'A', 'Mekhi Phifer', 2),
 ('High School High', 'United Artists Riverside (Galleria) Tyler Mall', '9:00 PM', 'A', 'Tia Carrera', 2),
 ('High School High', 'United Artists Riverside (Galleria) Tyler Mall', '9:00 PM', 'A', 'and Louise Fletcher', 2),
 ('High School High', 'United Artists Riverside (Galleria) Tyler Mall', '9:00 PM', 'D', 'Hart Bochner', 2)]

5. **Drop the `DATASET` table.**  
We don't need it anymore. 

6. **Save, commit, and push your work to GitHub.**  
Use the commit summary comment "Completed Part 3".